### PR TITLE
Viewer must be trusted for ssr templates to be supported

### PR DESCRIPTION
**Pull request by samouri**
**summary**
Viewer needs to be trusted for us to allow rendering templates server-side. This means that there are now three requirements to opt-in to ssr templating: doc-level opt-in, has capability, and being a trusted viewer.

**testing done**
- [x] new unit tests added for `isSupported`
- [ ] verify this doesn't break mail.ru / outlook

Note: I recommend viewing this PR [without whitespace](https://github.com/ampproject/amphtml/pull/25520/files?w=1).

Fixes b/142607068.

### CODE REVIEW

File: src/ssr-template-helper.js [hunk 1]
```diff
@@ -55,7 +55,7 @@ export class SsrTemplateHelper {
    */
   isSupported() {
     const ampdoc = this.viewer_.getAmpDoc();
-    if (ampdoc.isSingleDoc()) {
+    if (ampdoc.isSingleDoc() && this.viewer_.isTrustedViewer()) {
```
Comment (jridgewell) [inline src/ssr-template-helper.js]: `isTrustedViewer` returns a promise, not a boolean.

Comment (samouri) [reply src/ssr-template-helper.js]: 👍

**Comment by samouri** [comment]
I just found an issue with this I need to resolve before this can go through. I don't think `allowConsoleError` currently integrates well when the passed function is async (returns a promise).

**Comment by samouri** [comment]
In this presubmit logs: 
```
[19:24:10] Found forbidden: "isTrustedViewer" in src/ssr-template-helper.js:61:25
[19:24:10] Usage of this API requires dedicated review due to being privacy sensitive. Please file an issue asking for permission to use if you have not yet done so.
```

Should I still file a separate issue?

**Comment by choumx** [comment]
Nah, you can add it directly to the source whitelist in presubmit-checks.js.

**Comment by samouri** [comment]
I was able to pretty easily confirm that this doesn't break gmail by using charles proxy. I'm afraid this isn't as simple for email clients like `mail.ru` since they don't serve from the amp domain.

File: extensions/amp-form/0.1/amp-form.js [hunk 2]
```diff
@@ -1007,12 +1013,13 @@ export class AmpForm {
...
    */
   assertSsrTemplate_(value, msg) {
-    const supported = this.ssrTemplateHelper_.isSupported();
-    userAssert(
-      supported === value,
-      '[amp-form]: viewerRenderTemplate | %s',
-      msg
-    );
+    this.ssrTemplateHelper_.isSupported().then(supported => {
+      userAssert(
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: This has changed from a sync assert (a throw) to a possibly rejected promise. You need to wait on the value.

Comment (samouri) [reply extensions/amp-form/0.1/amp-form.js]: Yeah, I just spotted this as well. When I wrote this, I didn't know that `userAssert` throws. Working on it now 👍

Comment (samouri) [reply extensions/amp-form/0.1/amp-form.js]: meh, my solution stinks. I'll take another look tomorrow to see if I can make it cleaner.

**Comment by samouri** [comment]
The fix for `amp-form` is significantly more complex than the rest. I'll split it into its own PR.

**Comment by samouri** [comment]
Closing this in favor of https://github.com/ampproject/amphtml/pull/25638

**Pull request closed without merging**